pkg/wireguard/agent/agent.go

```diff
@@ -700,3 +763,61 @@ func (p *peerConfig) insertAllowedIP(ip net.IPNet) (updated bool) {
 	p.allowedIPs = append(p.allowedIPs, ip)
 	return true
 }
+
+// markRestoredIPsSeen removes the seen IPs from the list of restored allowed IPs.
+// Additionally, it cross-checks whether they were originally associated with the
+// expected peer, and removes them from the previous peer(s) otherwise.
+func (a *Agent) markRestoredIPsSeen(pubKey wgtypes.Key, seen ...net.IPNet) {
+	if len(a.restoredAllowedIPs) == 0 {
+		return
+	}
+
+	for _, ipn := range seen {
+		pfx := ipnetToPrefix(ipn)
+		prevKey, ok := a.restoredAllowedIPs[pfx]
+		delete(a.restoredAllowedIPs, pfx)
+
+		if !ok || prevKey == pubKey {
+			continue
+		}
+
+		// The restored allowed IP appears to be associated with a different
+		// peer. Hence, let's drop it from the other peer, to prevent flipping
+		// depending on which configuration gets applied last.
+		prevNode, ok := a.nodeNameByPubKey[prevKey]
+		if !ok {
+			continue
+		}
+
+		prevPeer, ok := a.peerByNodeName[prevNode]
+		if !ok {
+			continue
+		}
+
+		// It is not necessary that we push down this update, as WireGuard will
+		// automatically remove it from the old peer once it gets associated
+		// with the new one. However, we need to remove it from the local cache,
+		// to avoid reverting the change again upon a subsequent update.
+		prevPeer.removeAllowedIP(ipn)
+	}
+}
+
+func (a *Agent) computeRestoredIPs(pubKey wgtypes.Key) []net.IPNet {
```

Comment:
tiny nit: A short godoc comment what this does would be nice. Took me a second to understand what this does (it basically syncs the marks added by `markRestoredIPsSeen` over to `restoredPeers`).

Which makes me wonder: Is there a reason we can't make this logic here part of `markRestoredIPsSeen` and have it return the same list of currently stale IPs? 

Is it for performance that we do this "lazily" or is there a correctness thing I missed where doing it whenever we mark an IP would be wrong?

Reply:
> tiny nit: A short godoc comment what this does would be nice.

Yep, makes totally sense, added.

> Which makes me wonder: Is there a reason we can't make this logic here part of markRestoredIPsSeen and have it return the same list of currently stale IPs?

> Is it for performance that we do this "lazily" or is there a correctness thing I missed where doing it whenever we mark an IP would be wrong?

Essentially performance, to avoid having to iterate over the list of restored IPs and convert `net.IPNet` objects to `netip.Prefix` every time that we observe an ipcache change. This is based on the fact that, in the common case, `computeRestoredIPs` is expected to be called only twice for each peer, once during the initial update (in which the majority of IPs is expected to be restored), and once when restoration completes (when all IPs except stale ones need to be filtered). 

Another possibility could be storing the list of restored IPs of a given peer as a set to enable direct access/removals, but that would imply even more conversions to `netip.Prefix` (given that `net.IPNet` is not comparable), hence increasing the number of overall allocations.

Reply:
Great, thanks for the context!

Eventually, the whole WireGuard userspace state (not just the restorted part) might actually be a good fit to put into StateDB and a generic reconciler, allowing us to remove all of those maps. But that's a larger refactor not to be backported.

Reply:
Yep, makes sense. I tried to keep the changes to a minimum here for ease of backport, even though they still ended up not being that small in order to account for the different cases.